### PR TITLE
Updates because I tried to use this, also adding submodels

### PR DESCRIPTION
**Pull request by oxinabox**

### CODE REVIEW

File: src/Models.jl [hunk 1]
```diff
@@ -44,6 +45,26 @@ Returns a predictive distribution or point estimates depending on the [`Model`](
 """
 function predict end
 
+"""
+    submodels(model|template)
```
Comment (nickrobinson251) [inline src/Models.jl]:
```suggestion
    submodels(::Union{Template, Model})
```

File: src/Models.jl [hunk 2]
```diff
@@ -28,10 +28,11 @@ Defined as well are the traits:
 abstract type Model end
 
 """
-   fit(::Template, output, input) -> Model
+   fit(::Template, output, input, [weights]) -> Model
 
 Fit the [`Template`](@ref) to the `output` and `input` data and return a trained
 [`Model`](@ref).
+Convention is that `weights` defaults to `uweights(Float32, size(outputs, 2))``
```
Comment (nickrobinson251) [inline src/Models.jl]:
```suggestion
Convention is that `weights` defaults to `StatsBase.uweights(Float32, size(outputs, 2))`
```

File: src/Models.jl [hunk 3]
```diff
@@ -44,6 +45,26 @@ Returns a predictive distribution or point estimates depending on the [`Model`](
...
+
+Returns all submodels within a multistage model/template.
+Submodels are models within a model that have their own inputs (which may or may not be
+combined whith outputs of earlier models, before actually inputting them).
+Such multistage models take a tuple of inputs (which may be nested if the submodel itself
+has submodels).
+The order of inner models return by `submodels` is as per the order of the inputs in the
+tuple.
+
+For single-stage models, (i.e. ones that simply take a matrix as input), this returns a
```
Comment (nickrobinson251) [inline src/Models.jl]:
```suggestion
For single-stage models, (i.e. ones that simply take a matrix as input), this returns an
```

File: src/test_utils.jl [hunk 4]
```diff
@@ -122,94 +122,104 @@ end
 StatsBase.predict(m::FakeModel, inputs) = m.predictor(m.num_variates, inputs)
 
 """
-    test_interface(temp::Template; inputs=rand(5, 5), outputs=rand(5, 5))
+    test_interface(template::Template; inputs=rand(5, 5), outputs=rand(5, 5))
```
Comment (nickrobinson251) [inline src/test_utils.jl:125]: thank you

File: src/Models.jl [hunk 5]
```diff
@@ -44,6 +45,26 @@ Returns a predictive distribution or point estimates depending on the [`Model`](
 """
 function predict end
 
+"""
+    submodels(model|template)
+
+Returns all submodels within a multistage model/template.
```
Comment (nickrobinson251) [inline src/Models.jl]:
```suggestion
Return all submodels within a multistage model/template.

```

File: src/Models.jl [hunk 6]
```diff
@@ -44,6 +45,26 @@ Returns a predictive distribution or point estimates depending on the [`Model`](
...
+Submodels are models within a model that have their own inputs (which may or may not be
+combined whith outputs of earlier models, before actually inputting them).
```
Comment (nickrobinson251) [inline src/Models.jl]:
```suggestion
combined with outputs of _earlier_ submodels, before actually being passed as input to the submodel).
```

File: src/Models.jl [hunk 7]
```diff
@@ -44,6 +45,26 @@ Returns a predictive distribution or point estimates depending on the [`Model`](
...
+Such multistage models take a tuple of inputs (which may be nested if the submodel itself
+has submodels).
+The order of inner models return by `submodels` is as per the order of the inputs in the
```
Comment (nickrobinson251) [inline src/Models.jl]:
```suggestion
The order of submodels returned by `submodels` is as per the order of the inputs in the
```

File: src/Models.jl [hunk 8]
```diff
@@ -44,6 +45,26 @@ Returns a predictive distribution or point estimates depending on the [`Model`](
...
+tuple.
+
+For single-stage models, (i.e. ones that simply take a matrix as input), this returns a
+empty tuple.
+Wrapper models which do not expose their inner models to seperate inputs, including ones
```
Comment (nickrobinson251) [inline src/Models.jl:62]: i'd be tempted to make this a note
```
!!! note 
    Wrapper models which ...
```

**Review by nickrobinson251 — APPROVED** [review]
A few docs suggestions, otherwise looks good to me.

This PR touch much longer for me to review than it could have, because adding `submodels` was mixed in with a few other smaller changes, and i lost track of what was required for the `submodels` addition and what were unrelated changes.

File: src/Models.jl (hunk 8 above)
Comment (oxinabox) [reply src/Models.jl:62]: I am not.
Its not a side point or a remark not to be missed, it is a core part of the point

Comment (nickrobinson251) [reply src/Models.jl:62]: yeah, that's why i think it should be in an admonition: it's something not obvious and is worth calling out / drawing attention to. Up to you :)

Comment (oxinabox) [reply src/Models.jl:62]: I am fine for now, but may revisit this when we fully formalize the API for multistage models

**Pull request merged**